push_notifications/wns.py

```diff
@@ -0,0 +1,359 @@
+	if message:
+		wns_type = "wns/toast"
+		if isinstance(message, str):
+			message = {
+				"text": [message, ],
+			}
+		prepared_data = _wns_prepare_toast(data=message, **kwargs)
+	# Create a toast/tile/badge notification from a dictionary
+	elif xml_data:
+		xml = dict_to_xml_schema(xml_data)
+		wns_type = "wns/%s" % xml.tag
+		prepared_data = tostring(xml)
+	# Create a raw notification
+	elif raw_data:
+		wns_type = "wns/raw"
+		prepared_data = raw_data
+	else:
+		raise TypeError("At least one of the following parameters cannot be None type: "
+						"['message', 'xml_data', 'raw_data']")
+
+	_wns_send(uri=uri, data=prepared_data, wns_type=wns_type)
+
+
+def wns_send_bulk_message(uri_list, message=None, xml_data=None, raw_data=None, **kwargs):
+	"""
+	WNS doesn't support bulk notification, so we loop through each uri.
+
+	:param uri_list: list: A list of uris the notification will be sent to.
+	:param message: str: The notification data to be sent.
+	:param xml_data: dict: A dictionary containing data to be converted to an xml tree.
+	:param raw_data: str: Data to be sent via a 'raw' notification.
+	"""
+	if uri_list:
+		for uri in uri_list:
+			wns_send_message(uri=uri, message=message, xml_data=xml_data, raw_data=raw_data, **kwargs)
+
+
+def dict_to_xml_schema(data):
+	"""
+	Input a dictionary to be converted to xml. There should be only one key at the top level. The value must be a dict
+	with (required)'children' key and (optional)'attrs' key. This will be called the 'sub-element dictionary'.
+
+	The 'attrs' value must be a dictionary; each value will be added to the element's xml tag as attributes:
+
+		e.g.:
+			{
+				'example': {
+					'attrs': {
+						'key1': 'value1',
+						...
+					},
+					...
+				}
+			}
+
+		would result in:
+			<example key1="value1" key2="value2"></example>
+
+	The 'children' key is required.
+	If the value is a dict it must contain one or more keys which will be used as the sub-element names. Each
+	sub-element must have a value of a sub-element dictionary(see above) or a list of sub-element dictionaries.
+	If the value is not a dict, it will be the value of the element.
+	If the value is a list, multiple elements of the same tag will be created from each sub-element dict in the list.
+
+	:param data: dict: Used to create an XML tree.
+		e.g.:
+			example_data = {
+				'toast': {
+					'attrs': {
+						'launch': 'param',
+						'duration': 'short',
+					},
+					'children': {
+						'visual': {
+							'children': {
+								'binding': {
+									'attrs': {
+										'template': 'ToastText01',
+									},
+									'children': {
+										'text': [
+											{
+												'attrs': {
+													'id': '1',
+												},
+												'children': 'text1',
+											},
+											{
+												'attrs': {
+													'id': '2',
+												},
+												'children': 'text2',
+											},
+										],
+									},
+								},
+							},
+						},
+					},
+				},
+			}
```

Comment:
Alright, I've looked into it... I'm not sure how to format this correctly with tabs. What's your preference? Just put it all one line-ish?